src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/service/NonAssociationsService.kt

```diff
@@ -10,6 +10,7 @@ import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.dto.NonAssociation a
 import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.jpa.NonAssociation as NonAssociationJPA
 
 @Service
+@Transactional
```

Comment:
🤔 

I know this service doesn't do much but my idea was to not have a blanket `@Transactional` for all methods, and add it only for methods that needed it.

With that in mind, a method like `createNonAssociation()` would not be trasactional, it would use another method to persist the record and then publish a domain event (which we know would be problematic if done within a transaction as if for whatever reason there is a problem and the transaction is rolled back it whoever process it may get a 404)

Does this make sense? I would prefer to only have the `persistNonAssociation()` method to be transactional to avoid these kind of problems (and this is simpler than using more complicated things like using a separate class for messaging)

Reply:
Just to be clear, the actual sending of domain events would be in its own class, but invoking these would happen in the `NonAssociationsService` rather than having to wrap `NonAssociationsService` inside some other class just to avoid the transactional problems - if that make sense.

Happy to have a chat if that is not clear.

Reply:
That's why there was that [`.also {}` placeholder](https://github.com/ministryofjustice/hmpps-non-associations-api/pull/28/files#diff-46b4197c69d9454a21c6abbd82fb1663e6bc1f26f53cfa8695bbd77a8b4393f9L23-L25) in the service.

the `createNonAssociation()` method (non-transactional) would use `persistNonAssociation()` (transactional) _**and then**_ publish the domain event to notify other services (outside the transaction, once this is committed)

Reply:
Yes makes sense but I've wrapped the service call with the event /audit in the controller resource.